src/GovUk.Education.ExploreEducationStatistics.Common/Utils/StorageInstanceCreationUtil.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using GovUk.Education.ExploreEducationStatistics.Common.Utils.Interfaces;
+
+namespace GovUk.Education.ExploreEducationStatistics.Common.Utils;
+
+public class StorageInstanceCreationUtil : IStorageInstanceCreationUtil
+{
+    private static readonly ConcurrentDictionary<string, Unit> CreatedInstances = new();
+
+    public async Task CreateInstanceIfNotExistsAsync(
+        string connectionString,
+        AzureStorageType storageType,
+        string instance,
+        Func<Task> createIfNotExists)
+    {
+        var identifyingString = $"{instance}{storageType.ToString()}{connectionString}";
+        if (!CreatedInstances.ContainsKey(identifyingString))
+        {
+            await createIfNotExists.Invoke();
+            CreatedInstances[identifyingString] = Unit.Instance;
+        }
```

Comment:
It might not be as clear on the intent but just pointing out rather than `ContainsKey` followed by the add, both of these methods could be written in the style
```
if (CreatedInstances.TryAdd(identifyingString, Unit.Instance))
{
    await createIfNotExists.Invoke();
}
```

Reply:
Difficult one that - I wouldn't've immediately understood the TryAdd one, but would've twigged eventually.  Tempted to leave it.